### PR TITLE
Thumbnails on gallery's index #1771

### LINKED ISSUE #1771: Use thumbnails in gallery index
Gallery index is showing just a list of folder's name for each gallery. I don't like this, so I changed the behavior to build something better. An example here:

http://argentinaenpython.com.ar/galeria/

In the example you can see that I'm showing the title, preview image and a description. I'm using `.. previewimage:`, `.. description:` and `.. title:` from the `index.txt` file inside the gallery folder.

Here is part of the code that I'm building to make this work as in the example, but I need some help from you guys to make it perfect for a PR.

``` diff
diff --git a/nikola/conf.py.in b/nikola/conf.py.in
index d6776fd..54438a1 100644
--- a/nikola/conf.py.in
+++ b/nikola/conf.py.in
@@ -645,6 +645,19 @@ COMMENT_SYSTEM_ID = ${COMMENT_SYSTEM_ID}
 # Enable comments on picture gallery pages?
 # COMMENTS_IN_GALLERIES = False

+# Use a thumbnail (defined by ".. previewimage:") in list of galleries
+# for each gallery
+# GALLERIES_USE_THUMBNAIL = False
+
+# Columns of thumbnails list in galleries index
+# GALLERIES_COLUMNS = 3
+
+# Image to use as thumbnail for those galleries that doesn't have one
+# GALLERIES_DEFAULT_THUMBNAIL = ""
+
+# HTML title for /GALLERY_FOLDERS/index.html
+# GALLERY_INDEX_TITLE = "galleries"
+
 # What file should be used for directory indexes?
 # Defaults to index.html
 # Common other alternatives: default.html for IIS, index.php
diff --git a/nikola/data/themes/bootstrap3/templates/gallery.tmpl b/nikola/data/themes/bootstrap3/templates/gallery.tmpl
index 3d6c01a..5ae23e7 100644
--- a/nikola/data/themes/bootstrap3/templates/gallery.tmpl
+++ b/nikola/data/themes/bootstrap3/templates/gallery.tmpl
@@ -15,11 +15,40 @@
     </p>
     %endif
     %if folders:
-    <ul>
-    % for folder, ftitle in folders:
-        <li><a href="${folder}"><i class="glyphicon glyphicon-folder-open"></i>&nbsp;${ftitle}</a></li>
+    % if GALLERIES_USE_THUMBNAIL:
+    % for folder, ftitle, fpost in folders:
+   % if loop.index % GALLERIES_COLUMNS == 0:
+   <div class="row">
+   % endif
+     <div class="col-sm-6 col-md-4">
+       <div class="thumbnail">
+         <a href="${folder}">
+       % if fpost and getattr(fpost, 'previewimage', None):
+       <img src="${folder}${fpost.previewimage}" alt="${ftitle}">
+       % else:
+       <img src="${GALLERIES_DEFAULT_THUMBNAIL}" alt="${ftitle}">
+       % endif
+         </a>
+         <div class="caption">
+       <h3>${ftitle}</h3>
+       % if fpost and fpost.description():
+       <p>${fpost.description()}</p>
+       % endif
+         </div>
+       </div>
+     </div>
+   % if loop.index % GALLERIES_COLUMNS == (GALLERIES_COLUMNS - 1):
+   </div>
+   % endif
     % endfor
-    </ul>
+
+    % else:
+        <ul>
+            % for folder, ftitle in folders:
+                <li><a href="${folder}"><i class="glyphicon glyphicon-folder-open"></i>&nbsp;${ftitle}</a></li>
+            % endfor
+        </ul>
+    % endif
     %endif

 <div id="gallery_container"></div>
diff --git a/nikola/plugins/task/galleries.py b/nikola/plugins/task/galleries.py
index 39afd38..49807b4 100644
--- a/nikola/plugins/task/galleries.py
+++ b/nikola/plugins/task/galleries.py
@@ -237,7 +237,10 @@ class Galleries(Task, ImageProcessor):
                         ft = folder
                     if not folder.endswith('/'):
                         folder += '/'
-                    folders.append((folder, ft))
+                    if GALLERIES_USE_THUMBNAIL:
+                        folders.append((folder, ft, fpost))
+                    else:
+                        folders.append((folder, ft))

                 context["folders"] = natsort.natsorted(
                     folders, alg=natsort.ns.F | natsort.ns.IC)
```

It works without problem in my PC, but I want to make a PR so it can be included in Nikola to be available to other users also. The thing is that I don't know how to:
- get access to configuration variables from `gallery.tmpl`, like `GALLERIES_USE_THUMBNAIL`, for example
- change the `<title>` tag of /galleries/index.html and make it translatable (like other configs)
- where (and which one) I should store the default thumbnail image to show in those galleries that don't have one?
- should I add this to all the templates or just `bootstrap` and `bootstrap3`?
- how I define that /galleries/index.html depends on all the /galleries/<gallery-name>/index.txt? I need this because if I change the `.. previewimage:`, `.. title:` or something it should change

Thank you people!

### PR DESCRIPTION
**Pull request by humitos**
New settings to allow thumbnails in gallery's index.
Fixes #1771.

[Review on Reviewable](https://reviewable.io/reviews/getnikola/nikola/1779)

### CODE REVIEW

File: nikola/conf.py.in [hunk 1]
```diff
@@ -645,6 +645,21 @@ COMMENT_SYSTEM_ID = ${COMMENT_SYSTEM_ID}
 # Enable comments on picture gallery pages?
 # COMMENTS_IN_GALLERIES = False
 
+# Use a thumbnail (defined by ".. previewimage:") in list of galleries
+# for each gallery
+GALLERIES_USE_THUMBNAIL = True
+
+# Columns of thumbnails list in galleries index
+GALLERIES_COLUMNS = 3
```
Comment (ralsina) [inline nikola/conf.py.in]: if you use THUMBNAIL_SIZE as a guide for the sice of the thumbs, then this is probably not needed, just put them one after the other and as many as can fil will fit.

Keeping this setting means we have to ask themes to respect it and that's going to be difficult.

Comment (humitos) [inline nikola/conf.py.in]: How do you calculate "How many columns?" based on the THUMBNAIL_SIZE? I used that setting to put a `<div class="row">` or a `<div class="col-md-4">`.

I know *nothing* about Bootstrap, so if you know another better way to do that, please let me know.

Comment (ralsina) [inline nikola/conf.py.in]: I suppose using divs with display:inline will fit as many as possible per row then wrap. But hey, no HTML expert here :-)

**Comment by humitos** [comment]
Sorry, I did a new push by merging the latest tag into this branch.

I'm using this branch in my project at http://argentinaenpython.com/ , so from time to time I do a merge but I'm not sure if you like this.

Anyway, is this something still interesting for you to be part of nikola or not. In case the answer is "yes", please let me know what are the changes I should make to this PR. This is very old and I haven't read the new nikola source code since that, so probably there are many new things that I'm not aware.

Thanks!

**Comment by Kwpolska** [comment]
I think this is a good idea. But note that you’ll need to rebase this branch, or perhaps recreate it from scratch, based on `master` — you merged `v7-maintenance` which diverged from `master` a few months ago.

Delete your local copy of the branch, create new branch from master, and force-push. (If you do `git push --force origin gallery-thumbnails`, nobody will mind; make sure to include the branch in the command)

(Sorry, apparently nobody got around to reviewing this in 2 years. Happens sometimes.)

**Comment by humitos** [comment]
Hi! I rebased this PR onto `master`.

I saw that bootstrap3 is not more available, so I will need to test this a little more in bootstrap4. Maybe, in 3 years more we can merge it :laughing:

File: nikola/data/themes/bootstrap4/templates/gallery.tmpl [hunk 2]
```diff
@@ -15,11 +15,37 @@
     </p>
     %endif
     %if folders:
+    % if galleries_use_thumbnail:
+    <div class="row">
+    % for folder, ftitle, fpost in folders:
+	  <div class="col-md-4 col-sm-6 col-xs-12" style="min-height: 425px;">
+	    <div class="thumbnail">
```
Comment (Kwpolska) [inline nikola/data/themes/bootstrap4/templates/gallery.tmpl]: I’m afraid that doesn’t work in bs4. The new `card` classes would have to be used.

**Comment by humitos** [comment]
I used the `card` component from bootstrap4:

captura de pantalla_2018-04-22_20-23-19

I have no idea of CSS, so I tweak it a little to make it look "kind of pretty" but probably it need more work to make it more configurable and nicer.

I tested this by using this branch and running:

1. create nikola demo site
```
nikola init --demo --quiet demo
```
1. edit galleries/demo/index.txt to add these lines:
```
.. previewimage: tesla_lightning2_lg.jpg
.. description: Some public domain pictures of Nikola Tesla
```
1. Copy the default gallery
```
cd demo
cp -r galleries/demo galleries/demo1
...
```
1. Run `nikola auto -b`
1. Go to the galleries URL: http://127.0.0.1:8000/galleries/

File: nikola/data/themes/bootstrap4/templates/gallery.tmpl [hunk 3]
```diff
@@ -15,11 +15,36 @@
     </p>
     %endif
     %if folders:
+    % if galleries_use_thumbnail:
+    <% card_min_size = thumbnail_size + 100 %>
```
Comment (humitos) [inline nikola/data/themes/bootstrap4/templates/gallery.tmpl]: This should probably be a setting also: `GALLERIES_CARD_HEIGHT` and `GALLERY_CARD_WIDTH`

**Comment by ralsina** [comment]
Closed with thanks :-)

**Pull request closed without merging**